### PR TITLE
WebGPURenderer: Introduce `TiledLighting`

### PR DESCRIPTION
**Pull request by sunag**
### Description

The main idea of ​​having a lighting system is to open up the possibility for contributors to add their own lighting system like clustered lighting, which is why I added `TiledLighting` in addons.

About `TiledLighting`, there are still improvements to be made, such as sorting the lights according to the proximity of the camera to obtain better indexes, adding disposes, math optimizations and mainly added WebGLBackend support.

### API

The user does not need to manipulate anything special, such as shaders or uniforms, just define the class for light system.

```js
const lighting = new TiledLighting(); // ( maxLights = 1024, tileSize = 32 )

renderer = new THREE.WebGPURenderer();
renderer.lighting = lighting; // set lighting system
```

Or define per material.

```js
import { tiledLights } from 'three/addons/tsl/lighting/TiledLightsNode.js';

const lightsNode = tiledLights();
lightsNode.setLights( [ ...pointLights ] );

const materialA = new THREE.NodeMaterial();
materialA.lightsNode = lightsNode; 

const materialB = new THREE.NodeMaterial();
materialB.lightsNode = lightsNode; 
```

### Process summary

1. Store all the light data in a texture. It's position, color, distance...

2. Create a buffer that represents pieces of the image, the tiles. Each tile will have the indexes of the lights that are visible in that piece.

3. Instead of processing all the lights on the entire screen, we will only process the lights that are present in the indexes of each tile, I made each tile store 8 indexes. It is as if it had only 8 lights to calculate per pixel.

4. There is a pre-renderer stage where the compute-shader will project the position and radius of light to screen space where we can check for intersection with the tiles using AABB and thus bind the light, indexing it in the tile buffer.

_As I limited it to 8 indexes per tile and we do not have a sort lights, the system may fail if too many lights are added close to each other._

https://github.com/user-attachments/assets/2957adc7-bb5e-4bf3-adb3-3d21c1702051

### Commits

- [x] [introduce lighting system](https://github.com/mrdoob/three.js/commit/73b87746ae8f22683684376d642b8fc9e19693c7)
- [x] [introduce tiled lighting](https://github.com/mrdoob/three.js/commit/9b291ded2adb15359e35675695df54c296d0cab0)
- [x] [WGSLNodeBuilder: Remove incompatible types](https://github.com/mrdoob/three.js/commit/160db662d21244de9e70ca5c79934c282634d2d4)
- [x] [Renderer: move .nodes.library -> .library](https://github.com/mrdoob/three.js/commit/05c4cd22d4de1bd2d3fcf88d26b744f3e714eabd)

### CODE REVIEW

File: examples/jsm/lighting/TiledLighting.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import * as THREE from 'three';
```
Comment (Mugen87) [inline examples/jsm/lighting/TiledLighting.js:1]: Instead of `import *` how about this instead?
```js
import { Lighting } from 'three';
```

**Comment by RenaudRohlinger** [comment]
Am I the only one with 5fps on the `webgpu_lights_tiled` example when using 100 lights, and having both my Chrome and Chrome Canary completely crashing when using 1000?

Also the demo looks odd, maybe related to the recent changes on the transmission pass of the Renderer:
Screenshot 2024-10-22 at 15 14 02


/cc @sunag

**Comment by Mugen87** [comment]
On `dev` I have a black screen and see the following warning in Chrome:

> Error creating pipeline state Fragment function exceeds available stack space
    at InitializeImpl (../../third_party/dawn/src/dawn/native/metal/RenderPipelineMTL.mm:449)

**Comment by mrdoob** [comment]
It hangs Chrome here (M1 Pro) and eventually I get this in the console:

Screenshot 2024-10-22 at 5 43 57 PM

**Comment by sunag** [comment]
I ended up turning it off after an unsuccessful cleanup, here it is again
https://github.com/mrdoob/three.js/pull/29722

**Comment by mrdoob** [comment]
This is how it looks like now on a Pixel 8:

Screenshot_20241023-083019

**Comment by mrdoob** [comment]
Looks good on ChromeOS!

Screenshot 2024-10-23 08 57 46

**Comment by sunag** [comment]
> This is how it looks like now on a Pixel 8:

Could it be cache?

**Comment by mrdoob** [comment]
> > This is how it looks like now on a Pixel 8:
> 
> Could it be cache?

I think it was yeah.
]Just tried again and it works properly 🎉

**Comment by ligaofeng0901** [comment]
If a directional light is added into scene, some area that point light not covered lose directional light. I add a  directional light with red color, it appear like this below:
image

**Pull request merged**